request-manager/graph/graph.go

```diff
@@ -173,57 +173,72 @@ func (g *Graph) PrintDot() {
 
 // hasCycles returns true iff the graph has at least one cycle in it.
 func (g *Graph) hasCycles() bool {
-	seen := map[string]*Node{g.Source.Id: g.Source}
-	return g.hasCyclesHelper(seen, g.Source)
+	inPath := map[string]*Node{g.Source.Id: g.Source}
+	visited := map[string]*Node{}
```

Comment:
You could make all the `visited` maps `map[string]struct{}` to save a bit of space (since the stored *Node is never actually retrieved).

Reply:
Good point, I will fix that